submission/distributor.go

```diff
@@ -68,12 +74,68 @@ func (d *Distributor) Run(ctx context.Context) {
 	}
 }
 
-// refreshRoots requests roots from Logs and updates local copy if necessary.
+// refreshRoots requests roots from Logs and updates local copy.
 func (d *Distributor) refreshRoots(ctx context.Context) {
-	// TODO(Mercurrent) add implementation.
+	type RootsResult struct {
+		LogURL string
+		Roots  []*x509.Certificate
+		Err    error
+	}
+	ch := make(chan RootsResult, len(d.logClients))
+
+	rctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	for logURL, lc := range d.logClients {
+		go func(logURL string, lc client.AddLogClient) {
+			res := RootsResult{LogURL: logURL}
+
+			roots, err := lc.GetAcceptedRoots(rctx)
+			if err != nil {
+				res.Err = fmt.Errorf("%s: couldn't collect roots. %s", logURL, err)
+				ch <- res
+				return
+			}
+			for _, r := range roots {
+				parsed, err := x509.ParseCertificate(r.Data)
+				if err != nil {
+					res.Err = fmt.Errorf("%s: unable to parse root cert: %s", logURL, err)
```

Comment:
Might be helpful to include some identifying feature of the cert in this error message, e.g. hash or subject.

Reply:
hm, all we have at that moment is ASN1Cert struct, that wraps only []byte
There's no additional info to log.

Reply:
One drive-by comment for use of `x509.ParseCertificate()` -- our fork of `crypto/x509` has unusual error behaviour, where it can return both a parsed cert *and* an error.  So testing with `if x509.IsFatal(err) {` might be better here and elsewhere.

Reply:
Thanks, David. Updated.